### PR TITLE
Bug fix extract local function errors C#7 and below

### PR DESCRIPTION
**Pull request by akhera99**
Fixes: #55031

Opts to capture variables instead of creating new parameters in language versions below c#8 because they do not support static local functions.

* Need to fix the call to the new method

### CODE REVIEW

**Review by davidwengier — APPROVED** [review]

File: src/EditorFeatures/CSharpTest/CodeActions/ExtractMethod/ExtractLocalFunctionTests.cs [hunk 1]
```diff
@@ -4851,5 +4851,54 @@ static void L()
...
+    {
+        return "";
+    }
+
+    int j = 2;
+    [Test]
+    public void Test1()
+    {
+        const string NAME = ""SOMETEXT"";
+        {|Rename:NewMethod|}(NAME);
```
Comment (MaStr11) [inline src/EditorFeatures/CSharpTest/CodeActions/ExtractMethod/ExtractLocalFunctionTests.cs:4893]: Shouldn't this read:
```suggestion
        {|Rename:NewMethod|}();
```

File: src/EditorFeatures/CSharpTest/CodeActions/ExtractMethod/ExtractLocalFunctionTests.cs [hunk 2]
```diff
@@ -4851,5 +4851,54 @@ static void L()
...
+
+        void NewMethod()
+        {
+            Assert.AreEqual(string.Format(NAME, 0, 0), SomeOtherMethod(j));
+        }
+    }
+}";
+            await TestAsync(code, expected, TestOptions.Script.WithLanguageVersion(LanguageVersion.CSharp7), index: CodeActionIndex);
+        }
     }
```
Comment (MaStr11) [inline src/EditorFeatures/CSharpTest/CodeActions/ExtractMethod/ExtractLocalFunctionTests.cs:4854]: The bug comes up with other symbols too. Can you add tests for these cases:

* local const (this one is covered)
* method parameters
* locals
* don't capture linq range variables.

**Review by MaStr11 — COMMENTED** [review]
I started to work on this in #55135 and the proposed solution was to capture or to rename the parameters.

https://github.com/dotnet/roslyn/pull/55135#issuecomment-887632843

File: src/EditorFeatures/CSharpTest/CodeActions/ExtractMethod/ExtractLocalFunctionTests.cs (hunk 1 above)
Comment (akhera99) [reply src/EditorFeatures/CSharpTest/CodeActions/ExtractMethod/ExtractLocalFunctionTests.cs:4893]: Yep I am working on changing the new method call.

**Comment by MaStr11** [comment]
@akhera99 If you struggle to find a way to identify a linq range variable: I would suggest adding a property to [`VariableSymbol`](https://github.com/dotnet/roslyn/blob/1bf45deec8ac71398bff976cd9bc21ac0b113da7/src/Features/Core/Portable/ExtractMethod/MethodExtractor.VariableSymbol.cs#L22):

```CS
protected abstract class VariableSymbol
{
    public abstract bool CanBeCapturedByLocalFunction { get; }
}
```

This should return false in `QueryVariableSymbol` and `true` in any other case.

**Comment by akhera99** [comment]
> @akhera99 If you struggle to find a way to identify a linq range variable: I would suggest adding a property to [`VariableSymbol`](https://github.com/dotnet/roslyn/blob/1bf45deec8ac71398bff976cd9bc21ac0b113da7/src/Features/Core/Portable/ExtractMethod/MethodExtractor.VariableSymbol.cs#L22):
> 
> ```cs
> protected abstract class VariableSymbol
> {
>     public abstract bool CanBeCapturedByLocalFunction { get; }
> }
> ```
> 
> This should return false in `QueryVariableSymbol` and `true` in any other case.
> 
> You can expose `CanBeCapturedByLocalFunction` like so in `VariableInfo`:
> 
> ```cs
> protected class VariableInfo
> {
>             /// <summary>
>             /// Returns true, if the variable could be either passed as a parameter
>             /// to the new local function or the local function can capture the variable.
>             /// </summary>
>             public bool CanBeCapturedByLocalFunction
>                 => _variableSymbol.CanBeCapturedByLocalFunction;
> }
> ```

Thank you for your help. I was trying to avoid looking explicitly at the AnalyzerResult information, but that is definitely the easiest way to handle it.

**Pull request merged**